test/cri-containerd/clone_test.go

```diff
@@ -0,0 +1,404 @@
+func Test_CloneContainersWithClonedPodPool_WCOW(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client := newTestRuntimeClient(t)
+	nClones := 3
+
+	pullRequiredImages(t, []string{imageWindowsNanoserver})
+
+	// create template pod & container
+	templatePodID, templateContainerID := createTemplateContainer(ctx, t, client, getTemplatePodConfig("templatepod"), getTemplateContainerConfig("templatecontainer"))
+	defer removePodSandbox(t, client, ctx, templatePodID)
+	defer stopPodSandbox(t, client, ctx, templatePodID)
+	defer removeContainer(t, client, ctx, templateContainerID)
+	defer stopContainer(t, client, ctx, templateContainerID)
+
+	// create multiple pods
+	clonedPodIDs := []string{}
+	clonedSandboxRequests := []*runtime.RunPodSandboxRequest{}
+	for i := 0; i < nClones; i++ {
+		cloneSandboxRequest := getClonedPodConfig(i, templatePodID)
+		clonedPodID := runPodSandbox(t, client, ctx, cloneSandboxRequest)
+		clonedPodIDs = append(clonedPodIDs, clonedPodID)
+		clonedSandboxRequests = append(clonedSandboxRequests, cloneSandboxRequest)
+		defer removePodSandbox(t, client, ctx, clonedPodID)
+		defer stopPodSandbox(t, client, ctx, clonedPodID)
+	}
+
+	// create multiple clones
+	clonedContainers := []string{}
+	for i := 0; i < nClones; i++ {
+		cloneContainerRequest := getClonedContainerConfig(i, templateContainerID)
+
+		cloneContainerRequest.PodSandboxId = clonedPodIDs[i]
+		cloneContainerRequest.SandboxConfig = clonedSandboxRequests[i].Config
+		clonedContainerID := createContainer(t, client, ctx, cloneContainerRequest)
+		startContainer(t, client, ctx, clonedContainerID)
+
+		// cleanup
+		defer removeContainer(t, client, ctx, clonedContainerID)
+		defer stopContainer(t, client, ctx, clonedContainerID)
+
+		clonedContainers = append(clonedContainers, clonedContainerID)
+	}
+
+	for i := 0; i < nClones; i++ {
+		verifyContainerExec(ctx, t, client, clonedContainers[i])
+	}
+}
+
+func Test_ClonedContainerRunningAfterDeletingTemplate(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	client := newTestRuntimeClient(t)
+
+	pullRequiredImages(t, []string{imageWindowsNanoserver})
+
+	templatePodID, templateContainerID := createTemplateContainer(ctx, t, client, getTemplatePodConfig("templatepod"), getTemplateContainerConfig("templatecontainer"))
+
+	clonedPodID, clonedContainerID := createClonedContainer(ctx, t, client, templatePodID, templateContainerID, 1)
+	defer removePodSandbox(t, client, ctx, clonedPodID)
+	defer stopPodSandbox(t, client, ctx, clonedPodID)
+	defer removeContainer(t, client, ctx, clonedContainerID)
+	defer stopContainer(t, client, ctx, clonedContainerID)
+
+	stopPodSandbox(t, client, ctx, templatePodID)
+	removePodSandbox(t, client, ctx, templatePodID)
+
+	verifyContainerExec(ctx, t, client, clonedContainerID)
+
+}
+
+// A test to verify that multiple templats can be created and clones
+// can be made from each of them simultaneously.
```

Comment:
Nit: this isn't really "simultaneously", could we update the function to do that or otherwise update this comment?

Reply:
Good point. Initially this wasn't supposed to be a true `simultaneous` test but now I think I should modify the test to actually create templates and clones simultaneously. I have updated the test.